src/ai/backend/agent/agent.py

```diff
@@ -1701,6 +1704,80 @@ async def create_kernel(
+                                return resource_spec
+
+                            async with self.registry_lock:
+                                for kernel_id in self.kernel_registry.keys():
+                                    scratch_dir = (
+                                        self.local_config["container"]["scratch-root"]
+                                        / str(kernel_id)
+                                    ).resolve()
+                                    kernel_resource_spec = await self.loop.run_in_executor(
+                                        None,
+                                        _kernel_resource_spec_read,
+                                        scratch_dir / "config" / "resource.txt",
+                                    )
+                                    if kernel_resource_spec is None:
+                                        continue
+                                    else:
+                                        for i in kernel_resource_spec.slots.keys():
+                                            if i in occupied_slots_from_kernel:
+                                                occupied_slots_from_kernel[i] = Decimal(
+                                                    occupied_slots_from_kernel[i]
+                                                ) + Decimal(kernel_resource_spec.slots[i])
+                                            else:
+                                                occupied_slots_from_kernel[i] = Decimal(
+                                                    kernel_resource_spec.slots[i]
+                                                )
+
+                            log.error(
+                                "resource allocation failed (InsufficientResource)\n{0}\nwith"
+                                " context tag: {1}\nat allocating {2} {3}\ntotal allocatable"
+                                " amount:"
+                                " {4}\n-----------------------------------------------\n(before"
+                                " allocation) compute plugin alloc map:\n"
+                                " {5}\n-----------------------------------------------\n(error"
+                                " during allocation) occupied slots from compute plugin alloc"
+                                " map:\n{6}\n-----------------------------------------------\noccupied"
+                                " slots from kernel resources: {7}",
+                                e.args[0],
+                                e.args[1],
+                                e.args[2],
+                                e.args[3],
+                                e.args[4],
+                                pprint.pformat(original_plugin_alloc_map),
+                                pprint.pformat(plugin_alloc_map),
+                                occupied_slots_from_kernel,
+                            )
```

Comment:
Since `InsufficientResource` now has so many arguments, how about redefining it to use named arguments and fields with explicit type annotations?

Example:
```python
from dataclasses import dataclass

@dataclass
class CustomException(Exception):
    foo: str
    baz: int

    def __reduce__(self):  # make it pickle-safe
        return (self.__class__, (self.foo, self.baz))
```

Also, even when formatting with Black, the long-running multi-line strings could be split into more "semantic" lines for better readability, e.g.,
```suggestion
                            log.error(
                                "resource allocation failed (InsufficientResource)\n"
                                "{0}\n"
                                "with context tag: {1}\n"
                                "at allocating {2} {3}\n"
                                "total allocatable amount: {4}\n"
                                "-----------------------------------------------\n"
                                "(before allocation) compute plugin alloc map:\n"
                                " {5}\n"
                                "-----------------------------------------------\n"
                                "(error during allocation) occupied slots from "
                                "compute plugin alloc map:\n"
                                " {6}\n"
                                "-----------------------------------------------\n"
                                "occupied slots from kernel resources: {7}",
                                e.args[0],
                                e.args[1],
                                e.args[2],
                                e.args[3],
                                e.args[4],
                                pprint.pformat(original_plugin_alloc_map),
                                pprint.pformat(plugin_alloc_map),
                                occupied_slots_from_kernel,
                            )
```

Reply:
Thanks!
I modified InsufficientResource to have fields and type annotations.
And for readability, the black formattor was temporarily disabled.
plz check 👍